PR: https://github.com/odoo/odoo/pull/241368

From: 6a56bd10cec73eabd78b3ab7ba09fe4ab3789243
From: Youssef (abyo)
Date: 2026-04-19 08:57:09

Breaking data model changes scores: del:44 + add:0, change matches:
-    _inherit = 'account.move.line'
-    picking_type_id = fields.Many2one(
-    ship_later = fields.Boolean(string="Ship Later")
-    warehouse_id = fields.Many2one('stock.warehouse', default=_default_warehouse_id, ondelete='restrict')
-    route_id = fields.Many2one('stock.route', string="Spefic route for products delivered later.")
-    picking_policy = fields.Selection([
-    picking_ids = fields.One2many('stock.picking', 'pos_order_id')
-    picking_count = fields.Integer(compute='_compute_picking_count')
-    failed_pickings = fields.Boolean(compute='_compute_picking_count')
-    picking_type_id = fields.Many2one('stock.picking.type', related='session_id.config_id.picking_type_id', string="Operation Type", readonly=False)
-    stock_reference_ids = fields.Many2many('stock.reference', 'stock_reference_pos_order_rel', 'pos_order_id', 'reference_id', string="Reference")
-    shipping_date = fields.Date('Shipping Date')
-    _inherit = ['pos.load.mixin']
-    pos_order_line_id = fields.Many2one('pos.order.line', index='btree_not_null')
-    order_id = fields.Many2one('pos.order', related="pos_order_line_id.order_id", readonly=False)
-    lot_name = fields.Char('Lot Name')
-    product_id = fields.Many2one('product.product', related='pos_order_line_id.product_id', readonly=False)
-    pack_lot_ids = fields.One2many('pos.pack.operation.lot', 'pos_order_line_id', string='Lot/serial Number')
-    failed_pickings = fields.Boolean(compute='_compute_picking_count')
-    picking_count = fields.Integer(compute='_compute_picking_count')
-    picking_ids = fields.One2many('stock.picking', 'pos_session_id')
-    update_stock_at_closing = fields.Boolean('Stock should be updated at closing')
-    _inherit = 'stock.rule'
-    point_of_sale_update_stock_quantities = fields.Selection([
-    update_stock_quantities = fields.Selection(related="company_id.point_of_sale_update_stock_quantities", readonly=False)
-    pos_picking_policy = fields.Selection(related='pos_config_id.picking_policy', readonly=False)
-    pos_picking_type_id = fields.Many2one(related='pos_config_id.picking_type_id', readonly=False)
-    pos_route_id = fields.Many2one(related='pos_config_id.route_id', readonly=False)
-    pos_ship_later = fields.Boolean(related='pos_config_id.ship_later', readonly=False)
-    pos_warehouse_id = fields.Many2one(related='pos_config_id.warehouse_id', readonly=False, string="Warehouse (PoS)")
-    _inherit = 'stock.picking'
-    pos_session_id = fields.Many2one('pos.session', index=True)
-    pos_order_id = fields.Many2one('pos.order', index=True)
-    _inherit = ['stock.picking.type', 'pos.load.mixin']
-    _inherit = 'stock.move'
-    _inherit = 'stock.reference'
-    pos_order_ids = fields.Many2many(
-    _inherit = "stock.warehouse"
-    pos_type_id = fields.Many2one('stock.picking.type', string="Point of Sale Operation Type", copy=False)

Total Changes: 1070

[IMP] point_of_sale: use POS without stock module

Point of sale module depends on the stock module, where
most complex operations take place. Some small businesses don't need
complex operations like batch picking, tracking products, etc..
They need simple operations that increases and decreases the product's
quantities in their inventory.

This commit seperates stock and POS, and adds a bridge
module pos_stock that will override the simple stock operations
(increase/decrease inventory) in the POS with the
actual complex stock operations in the stock module.

All stock operations currently in PoS will be moved to
pos_stock.

Following the same rationale, we allow pos_sale to be used solely
without the need of the stock module.
Not every business needs to create a picking/delivery
or track its products.

All stock operations currently in pos_sale will be
moved to pos_sale_stock.

closes odoo/odoo#241368

Task: 5418116
Related: odoo/enterprise#106117
Related: odoo/upgrade#9418
Signed-off-by: William André (wan) <wan@odoo.com>

================================= pseudo patch: =================================

--- a/addons/point_of_sale/models/__init__.py
+++ b/addons/point_of_sale/models/__init__.py
@@ -32,9 +32,6 @@ from . import res_company
 from . import res_config_settings
 from . import ir_http
 from . import ir_module_module
-from . import stock_picking
-from . import stock_reference
-from . import stock_warehouse
 from . import pos_payment
 from . import pos_payment_method
 from . import pos_bill

--- a/addons/point_of_sale/models/account_move.py
+++ b/addons/point_of_sale/models/account_move.py
@@ -33,32 +33,6 @@ class AccountMove(models.Model):
             if move.pos_session_ids:
                 move.always_tax_exigible = True
 
-
-    def _get_invoiced_lot_values(self):
-        self.ensure_one()
-
-        lot_values = super(AccountMove, self)._get_invoiced_lot_values()
-
-        if self.state == 'draft':
-            return lot_values
-
-        # user may not have access to POS orders, but it's ok if they have
-        # access to the invoice
-        for order in self.sudo().pos_order_ids:
-            for line in order.lines:
-                lots = line.pack_lot_ids or False
-                if lots:
-                    for lot in lots:
-                        lot_values.append({
-                            'product_name': lot.product_id.name,
-                            'quantity': line.qty if lot.product_id.tracking == 'lot' else 1.0,
-                            'uom_name': line.product_uom_id.name,
-                            'lot_name': lot.lot_name,
-                            'pos_lot_id': lot.id,
-                        } | self._extract_extra_invoiced_lot_values(lot))
-
-        return lot_values
-
     def _prepare_payments_widget_reconciled_info(self, partial_info):
         # EXTENDS account.
         results = super()._prepare_payments_widget_reconciled_info(partial_info)
@@ -112,20 +86,3 @@ class AccountMove(models.Model):
     @api.model
     def _load_pos_data_domain(self, data, config):
         return False
-
-class AccountMoveLine(models.Model):
-    _inherit = 'account.move.line'
-
-    def _get_cogs_value(self):
-        self.ensure_one()
-        if not self.product_id:
-            return self.price_unit
-        price_unit = super()._get_cogs_value()
-        sudo_order = self.move_id.sudo().pos_order_ids
-        if sudo_order:
-            price_unit = sudo_order._get_pos_anglo_saxon_price_unit(self.product_id, self.quantity)
-        return price_unit
-
-    def _compute_name(self):
-        amls = self.filtered(lambda l: not l.move_id.pos_session_ids)
-        super(AccountMoveLine, amls)._compute_name()

--- a/addons/point_of_sale/models/pos_config.py
+++ b/addons/point_of_sale/models/pos_config.py
@@ -26,12 +26,6 @@ class PosConfig(models.Model):
     _description = 'Point of Sale Configuration'
     _check_company_auto = True
 
-    def _default_warehouse_id(self):
-        return self.env['stock.warehouse'].search(self.env['stock.warehouse']._check_company_domain(self.env.company), limit=1).id
-
-    def _default_picking_type_id(self):
-        return self.env['stock.warehouse'].search(self.env['stock.warehouse']._check_company_domain(self.env.company), limit=1).pos_type_id.id
-
     def _default_sale_journal(self):
         journal = self.env['account.journal']._ensure_company_account_journal()
         return journal
@@ -79,13 +73,6 @@ class PosConfig(models.Model):
     use_order_printer = fields.Boolean('Order Printer')
     is_installed_account_accountant = fields.Boolean(string="Is the Full Accounting Installed",
         compute="_compute_is_installed_account_accountant")
-    picking_type_id = fields.Many2one(
-        'stock.picking.type',
-        string='Operation Type',
-        default=_default_picking_type_id,
-        required=True,
-        domain=lambda self: [('code', '=', 'outgoing'), ('warehouse_id.company_id', '=', self.env.company.id)],
-        ondelete='restrict')
     journal_id = fields.Many2one(
         'account.journal', string='Point of Sale Journal',
         domain=[('type', 'in', ('general', 'sale'))],
@@ -188,15 +175,6 @@ class PosConfig(models.Model):
     only_round_cash_method = fields.Boolean(string="Only apply rounding on cash")
     has_active_session = fields.Boolean(compute='_compute_current_session')
     manual_discount = fields.Boolean(string="Line Discounts", default=True)
-    ship_later = fields.Boolean(string="Ship Later")
-    warehouse_id = fields.Many2one('stock.warehouse', default=_default_warehouse_id, ondelete='restrict')
-    route_id = fields.Many2one('stock.route', string="Spefic route for products delivered later.")
-    picking_policy = fields.Selection([
-        ('direct', 'As soon as possible, with back orders'),
-        ('one', 'When all products are ready')],
-        string='Shipping Policy', required=True, default='direct',
-        help="If you deliver all products at once, the delivery order will be scheduled based on the greatest "
-        "product lead time. Otherwise, it will be based on the shortest.")
     auto_validate_electronic_payment = fields.Boolean(default=True, help="Automatically validates orders paid with an electronic payment (terminals, qr codes, ...).")
     trusted_config_ids = fields.Many2many("pos.config", relation="pos_config_trust_relation", column1="is_trusting",
                                           column2="is_trusted", string="Trusted Point of Sale Configurations",
@@ -622,11 +600,6 @@ class PosConfig(models.Model):
 
     @api.model_create_multi
     def create(self, vals_list):
-        if not self._default_warehouse_id():
-            self.env['stock.warehouse'].create({
-                'code': vals_list[0].get('name')[:3],  # first 3 characters of pos.config name
-                'company_id': self.env.company.id,
-            })
         for vals in vals_list:
             if not vals.get('iface_tipproduct', False):
                 vals['tip_product_id'] = False

--- a/addons/point_of_sale/models/pos_order.py
+++ b/addons/point_of_sale/models/pos_order.py
@@ -9,7 +9,7 @@ from pprint import pformat
 
 from markupsafe import Markup
 
-from odoo import api, fields, models, tools, _
+from odoo import api, fields, models, _
 from odoo.tools import float_is_zero, float_round, float_repr, float_compare, formatLang
 from odoo.exceptions import ValidationError, UserError
 from odoo.fields import Command, Domain
@@ -125,7 +125,7 @@ class PosOrder(models.Model):
         self.ensure_one()
         if not draft and self.state != 'cancel':
             self.action_pos_order_paid()
-            self._create_order_picking()
+            self._set_product_qty_available()
             self._compute_total_cost_in_real_time()
 
         self._generate_order_invoice()
@@ -175,10 +175,6 @@ class PosOrder(models.Model):
         if order.get('partner_id'):
             self.update_order_partner(order)
             existing_order.write({'partner_id': order.get('partner_id'), 'to_invoice': order.get('to_invoice', False)})
-        # update pickings
-        if order.get('shipping_date'):
-            existing_order.write({'shipping_date': order.get('shipping_date')})
-            existing_order.picking_ids.filtered(lambda p: p.state not in ['done', 'cancel']).write({'scheduled_date': order.get('shipping_date')})
         # payments
         if order.get('payment_ids'):
             self._update_lines(order, existing_order, ['lines', 'payment_ids'])
@@ -307,12 +303,6 @@ class PosOrder(models.Model):
                 }))
         return invoice_lines
 
-    def _get_pos_anglo_saxon_price_unit(self, product, quantity):
-        moves = self.mapped('picking_ids.move_ids')\
-            .filtered(lambda m: m.is_valued and m.product_id.valuation == 'real_time' and m.product_id.id == product.id)\
-            .sorted(lambda x: x.date)
-        return moves._get_price_unit()
-
     name = fields.Char(string='Order Ref', required=True, readonly=True, copy=False, default='/')
     last_order_preparation_change = fields.Char(string='Last preparation change', help="Last printed state of the order")
     date_order = fields.Datetime(string='Date', readonly=True, index=True, default=fields.Datetime.now)
@@ -349,11 +339,6 @@ class PosOrder(models.Model):
         'Status', readonly=True, copy=False, default='draft', index=True)
 
     account_move = fields.Many2one('account.move', string='Invoice', readonly=True, copy=False, index="btree_not_null")
-    picking_ids = fields.One2many('stock.picking', 'pos_order_id')
-    picking_count = fields.Integer(compute='_compute_picking_count')
-    failed_pickings = fields.Boolean(compute='_compute_picking_count')
-    picking_type_id = fields.Many2one('stock.picking.type', related='session_id.config_id.picking_type_id', string="Operation Type", readonly=False)
-    stock_reference_ids = fields.Many2many('stock.reference', 'stock_reference_pos_order_rel', 'pos_order_id', 'reference_id', string="Reference")
     preset_id = fields.Many2one('pos.preset', string='Preset')
     floating_order_name = fields.Char(string='Order Name')
     general_customer_note = fields.Text(string='General Customer Note')
@@ -368,7 +353,6 @@ class PosOrder(models.Model):
     payment_ids = fields.One2many('pos.payment', 'pos_order_id', string='Payments')
     session_move_id = fields.Many2one('account.move', string='Session Journal Entry', related='session_id.move_id', readonly=True, copy=False)
     to_invoice = fields.Boolean('To invoice', copy=False)
-    shipping_date = fields.Date('Shipping Date')
     preset_time = fields.Datetime(string='Hour', help="Hour of the day for the order")
     is_invoiced = fields.Boolean('Is Invoiced', compute='_compute_is_invoiced')
     is_tipped = fields.Boolean('Is this already tipped?', readonly=True)
@@ -458,12 +442,6 @@ class PosOrder(models.Model):
         for order in self:
             order.is_invoiced = bool(order.account_move)
 
-    @api.depends('picking_ids', 'picking_ids.state')
-    def _compute_picking_count(self):
-        for order in self:
-            order.picking_count = len(order.picking_ids)
-            order.failed_pickings = bool(order.picking_ids.filtered(lambda p: p.state != 'done'))
-
     @api.depends('date_order', 'company_id', 'currency_id', 'company_id.currency_id')
     def _compute_currency_rate(self):
         for order in self:
@@ -488,21 +466,12 @@ class PosOrder(models.Model):
         then the margin of said order is not computed (it will be computed when closing the session).
         """
         for order in self:
-            lines = order.lines
-            if not order._should_create_picking_real_time():
-                storable_fifo_avco_lines = lines.filtered(lambda l: l._is_product_storable_fifo_avco())
-                lines -= storable_fifo_avco_lines
-            stock_moves = order.picking_ids.move_ids
-            lines._compute_total_cost(stock_moves)
-
-    def _compute_total_cost_at_session_closing(self, stock_moves):
-        """
-        Compute the margin at the end of the session. This method should be called to compute the remaining lines margin
-        containing a storable product with a fifo/avco cost method and then compute the order margin
-        """
-        for order in self:
-            storable_fifo_avco_lines = order.lines.filtered(lambda l: l._is_product_storable_fifo_avco())
-            storable_fifo_avco_lines._compute_total_cost(stock_moves)
+            lines = order._get_total_cost_in_real_time_lines()
+            lines._compute_total_cost()
+
+    def _get_total_cost_in_real_time_lines(self):
+        self.ensure_one()
+        return self.lines
 
     @api.depends('lines.margin', 'is_total_cost_computed')
     def _compute_margin(self):
@@ -721,14 +690,6 @@ class PosOrder(models.Model):
     def get_reference_last_part(self):
         return self.pos_reference.split('-')[-1]
 
-    def action_stock_picking(self):
-        self.ensure_one()
-        action = self.env['ir.actions.act_window']._for_xml_id('stock.action_picking_tree_ready')
-        action['display_name'] = _('Pickings')
-        action['context'] = {}
-        action['domain'] = [('id', 'in', self.picking_ids.ids)]
-        return action
-
     def action_view_invoice(self):
         invoices = self.account_move
         if (len(invoices) == 1):
@@ -1048,33 +1009,6 @@ class PosOrder(models.Model):
                         'balance': balance,
                         'display_type': 'rounding',
                     })
-        # Stock.
-        if self.picking_ids.ids:
-            stock_moves = self.env['stock.move'].sudo().search([
-                ('picking_id', 'in', self.picking_ids.ids),
-                ('product_id.valuation', '=', 'real_time'),
-            ])
-            for stock_move in stock_moves:
-                product_accounts = stock_move.with_company(stock_move.company_id).product_id._get_product_accounts()
-                expense_account = product_accounts['expense']
-                stock_account = product_accounts['stock_valuation']
-                balance = stock_move.value if stock_move.is_out else -stock_move.value
-                aml_vals_list_per_nature['stock'].append({
-                    'name': _("Stock variation for %s", stock_move.product_id.name),
-                    'account_id': expense_account.id,
-                    'partner_id': commercial_partner.id,
-                    'currency_id': self.company_id.currency_id.id,
-                    'amount_currency': balance,
-                    'balance': balance,
-                })
-                aml_vals_list_per_nature['stock'].append({
-                    'name': _("Stock variation for %s", stock_move.product_id.name),
-                    'account_id': stock_account.id,
-                    'partner_id': commercial_partner.id,
-                    'currency_id': self.company_id.currency_id.id,
-                    'amount_currency': -balance,
-                    'balance': -balance,
-                })
 
         # sort self.payment_ids by is_split_transaction:
         for payment_id in self.payment_ids:
@@ -1171,11 +1105,7 @@ class PosOrder(models.Model):
     def action_pos_order_invoice(self):
         self.ensure_one()
         if not (move := self.account_move):
-            is_picking_created = self._should_create_picking_real_time()
-            self.write({'to_invoice': True})
-            if not is_picking_created and self._should_create_picking_real_time() and self.session_id.state != 'closed':
-                self._create_order_picking()
-            move = self._generate_pos_order_invoice()
+            move = self._prepare_missing_invoice_moves()
         return {
             'name': _('Customer Invoice'),
             'view_mode': 'form',
@@ -1187,6 +1117,10 @@ class PosOrder(models.Model):
             'res_id': move.id,
         }
 
+    def _prepare_missing_invoice_moves(self):
+        self.write({'to_invoice': True})
+        return self._generate_pos_order_invoice()
+
     def _get_invoice_post_context(self):
         return {"skip_invoice_sync": True}
 
@@ -1337,7 +1271,6 @@ class PosOrder(models.Model):
             'pos.session': [],
             'pos.payment': self.env['pos.payment']._load_pos_data_read(self.payment_ids, config) if config else [],
             'pos.order.line': self.env['pos.order.line']._load_pos_data_read(self.lines, config) if config else [],
-            'pos.pack.operation.lot': self.env['pos.pack.operation.lot']._load_pos_data_read(self.lines.pack_lot_ids, config) if config else [],
             'product.attribute.custom.value': self.env['product.attribute.custom.value']._load_pos_data_read(self.lines.custom_attribute_value_ids, config) if config else [],
             'account.move': self.env['account.move'].sudo()._load_pos_data_read(account_moves, config) if config else [],
         }
@@ -1347,31 +1280,6 @@ class PosOrder(models.Model):
         refunded_orderline_ids = [line[2]['refunded_orderline_id'] for line in order['lines'] if line[0] in [0, 1] and line[2].get('refunded_orderline_id')]
         return self.env['pos.order.line'].browse(refunded_orderline_ids).mapped('order_id')
 
-    def _should_create_picking_real_time(self):
-        return not self.session_id.update_stock_at_closing or self._force_create_picking_real_time()
-
-    def _force_create_picking_real_time(self):
-        return self.company_id.anglo_saxon_accounting and self.to_invoice
-
-    def _create_order_picking(self):
-        self.ensure_one()
-        if self.picking_ids:
-            return
-        if self.shipping_date:
-            self.sudo().lines._launch_stock_rule_from_pos_order_lines()
-        else:
-            if self._should_create_picking_real_time():
-                picking_type = self.config_id.picking_type_id
-                if self.partner_id.property_stock_customer:
-                    destination_id = self.partner_id.property_stock_customer.id
-                elif not picking_type or not picking_type.default_location_dest_id:
-                    destination_id = self.env['stock.warehouse']._get_partner_locations()[0].id
-                else:
-                    destination_id = picking_type.default_location_dest_id.id
-
-                pickings = self.env['stock.picking']._create_picking_from_pos_order_lines(destination_id, self.lines, picking_type, self.partner_id)
-                pickings.write({'pos_session_id': self.session_id.id, 'pos_order_id': self.id, 'origin': self.name})
-
     def add_payment(self, data):
         """Create a new payment for the order"""
         self.ensure_one()
@@ -1429,12 +1337,10 @@ class PosOrder(models.Model):
             refund_order = order.copy(
                 order._prepare_refund_values(current_session)
             )
-            for line in order.lines.filtered(lambda l: l.refunded_qty < l.qty):
-                PosPackOperationLot = self.env['pos.pack.operation.lot']
-                for pack_lot in line.pack_lot_ids:
-                    PosPackOperationLot += pack_lot.copy()
-                refund_line = line.copy(line._prepare_refund_data(refund_order, PosPackOperationLot))
-                refund_line._onchange_amount_line_all()
+            for line in order.lines:
+                if line.refunded_qty < line.qty:
+                    refund_line = line.copy(line._prepare_refund_data(refund_order))
+                    refund_line._onchange_amount_line_all()
             refund_order._compute_prices()
             refund_orders |= refund_order
             refund_order.config_id.notify_synchronisation(current_session.id, 0)
@@ -1575,25 +1481,18 @@ class PosOrder(models.Model):
     def _prepare_pos_log(self, body):
         return body
 
-
-class PosPackOperationLot(models.Model):
-    _name = 'pos.pack.operation.lot'
-    _description = "Specify product lot/serial number in pos order line"
-    _rec_name = "lot_name"
-    _inherit = ['pos.load.mixin']
-
-    pos_order_line_id = fields.Many2one('pos.order.line', index='btree_not_null')
-    order_id = fields.Many2one('pos.order', related="pos_order_line_id.order_id", readonly=False)
-    lot_name = fields.Char('Lot Name')
-    product_id = fields.Many2one('product.product', related='pos_order_line_id.product_id', readonly=False)
-
-    @api.model
-    def _load_pos_data_domain(self, data, config):
-        return [('pos_order_line_id', 'in', [line['id'] for line in data['pos.order.line']])]
+    def _set_product_qty_available(self):
+        if not self._should_update_quantity_on_product():
+            return
+        for line in self.lines:
+            if line.product_id.is_storable:
+                line.product_id.sudo().qty_available -= line.qty
 
     @api.model
-    def _load_pos_data_fields(self, config):
-        return ['lot_name', 'pos_order_line_id', 'write_date']
+    def _should_update_quantity_on_product(self):
+        """ Overriden in pos_stock, as the quantity update is handled through the stock module.
+        """
+        return True
 
 
 class AccountCashRounding(models.Model):

--- a/addons/point_of_sale/models/pos_order_line.py
+++ b/addons/point_of_sale/models/pos_order_line.py
@@ -1,7 +1,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from uuid import uuid4
-from datetime import UTC
 from markupsafe import Markup
 
 from odoo import api, Command, fields, models, _
@@ -44,7 +43,6 @@ class PosOrderLine(models.Model):
     order_id = fields.Many2one('pos.order', string='Order Ref', ondelete='cascade', required=True, index=True)
     tax_ids = fields.Many2many('account.tax', string='Taxes', readonly=True)
     tax_ids_after_fiscal_position = fields.Many2many('account.tax', compute='_get_tax_ids_after_fiscal_position', string='Taxes to Apply')
-    pack_lot_ids = fields.One2many('pos.pack.operation.lot', 'pos_order_line_id', string='Lot/serial Number')
     product_uom_id = fields.Many2one('uom.uom', string='Product Unit', related='product_id.uom_id')
     currency_id = fields.Many2one('res.currency', related='order_id.currency_id')
     full_product_name = fields.Char('Full Product Name')
@@ -74,7 +72,7 @@ class PosOrderLine(models.Model):
         return [
             'qty', 'attribute_value_ids', 'custom_attribute_value_ids', 'price_unit',
             'uuid', 'price_subtotal', 'price_subtotal_incl', 'order_id', 'note', 'price_type',
-            'product_id', 'discount', 'tax_ids', 'pack_lot_ids', 'customer_note',
+            'product_id', 'discount', 'tax_ids', 'customer_note',
             'refunded_qty', 'price_extra', 'full_product_name', 'refunded_orderline_id',
             'combo_parent_id', 'combo_line_ids', 'combo_item_id', 'refund_orderline_ids',
             'extra_tax_data', 'write_date',
@@ -86,16 +84,13 @@ class PosOrderLine(models.Model):
             refund_order_line = orderline.refund_orderline_ids.filtered(lambda l: l.order_id.state != 'cancel')
             orderline.refunded_qty = -sum(refund_order_line.mapped('qty'))
 
-    def _prepare_refund_data(self, refund_order, PosPackOperationLot):
+    def _prepare_refund_data(self, refund_order):
         """
         This prepares data for refund order line. Inheritance may inject more data here
 
         @param refund_order: the pre-created refund order
         @type refund_order: pos.order
 
-        @param PosPackOperationLot: the pre-created Pack operation Lot
-        @type PosPackOperationLot: pos.pack.operation.lot
-
         @return: dictionary of data which is for creating a refund order line from the original line
         @rtype: dict
         """
@@ -104,7 +99,6 @@ class PosOrderLine(models.Model):
             'name': _('%(name)s REFUND', name=self.name),
             'qty': -(self.qty - self.refunded_qty),
             'order_id': refund_order.id,
-            'pack_lot_ids': PosPackOperationLot,
             'is_total_cost_computed': False,
             'refunded_orderline_id': self.id,
         }
@@ -124,11 +118,6 @@ class PosOrderLine(models.Model):
         return super().create(vals_list)
 
     def write(self, vals):
-        if vals.get('pack_lot_line_ids'):
-            for pl in vals.get('pack_lot_ids'):
-                if pl[2].get('server_id'):
-                    pl[2]['id'] = pl[2]['server_id']
-                    del pl[2]['server_id']
         if self.order_id.config_id.order_edit_tracking and vals.get('qty') is not None and vals.get('qty') < self.qty:
             self.is_edited = True
             body = _("%(product_name)s: Ordered quantity: %(old_qty)s", product_name=self.full_product_name, old_qty=self.qty)
@@ -137,48 +126,6 @@ class PosOrderLine(models.Model):
                 line.order_id.message_post(body=line.order_id._prepare_pos_log(body))
         return super().write(vals)
 
-    @api.model
-    def get_existing_lots(self, company_id, config_id, product_id):
-        """
-        Return the lots that are still available in the given company.
-        The lot is available if its quantity in the corresponding stock_quant and pos stock location is > 0.
-        """
-        self.check_access('read')
-        pos_config = self.env['pos.config'].browse(config_id)
-        if not pos_config:
-            raise UserError(_('No PoS configuration found'))
-
-        src_loc = pos_config.picking_type_id.default_location_src_id
-
-        domain = [
-            '|',
-            ('company_id', '=', False),
-            ('company_id', '=', company_id),
-            ('product_id', '=', product_id),
-            ('location_id', 'in', src_loc.child_internal_location_ids.ids),
-            ('quantity', '>', 0),
-            ('lot_id', '!=', False),
-        ]
-
-        groups = self.sudo().env['stock.quant']._read_group(
-            domain=domain,
-            groupby=['lot_id'],
-            aggregates=['quantity:sum']
-        )
-
-        result = []
-        has_lot_expiration_date = 'expiration_date' in self.env['stock.lot']._fields
-        for lot_recordset, total_quantity in groups:
-            if lot_recordset:
-                result.append({
-                    'id': lot_recordset.id,
-                    'name': lot_recordset.name,
-                    'product_qty': total_quantity,
-                    'expiration_date': lot_recordset.expiration_date if has_lot_expiration_date else False,
-                })
-
-        return result
-
     @api.ondelete(at_uninstall=False)
     def _unlink_except_order_state(self):
         if self.filtered(lambda x: x.order_id.state not in ["draft", "cancel"]):
@@ -239,102 +186,16 @@ class PosOrderLine(models.Model):
             'pos_order_ids': [Command.link(self.order_id.id)],
         }
 
-    def _prepare_procurement_values(self):
-        """ Prepare specific key for moves or other components that will be created from a stock rule
-        coming from a sale order line. This method could be override in order to add other custom key that could
-        be used in move/po creation.
-        """
-        self.ensure_one()
-        # Use the delivery date if there is else use date_order and lead time
-        if self.order_id.shipping_date:
-            # get timezone from user
-            # and convert to UTC to avoid any timezone issue
-            # because shipping_date is date and date_planned is datetime
-            from_zone = self.env.tz
-            shipping_date = fields.Datetime.to_datetime(self.order_id.shipping_date)
-            shipping_date = shipping_date.replace(tzinfo=from_zone)
-            date_deadline = shipping_date.astimezone(UTC).replace(tzinfo=None)
-        else:
-            date_deadline = self.order_id.date_order
-
-        values = {
-            'date_planned': date_deadline,
-            'date_deadline': date_deadline,
-            'route_ids': self.order_id.config_id.route_id,
-            'warehouse_id': self.order_id.config_id.warehouse_id or False,
-            'partner_id': self.order_id.partner_id.id,
-            'product_description_variants': self.full_product_name,
-            'company_id': self.order_id.company_id,
-            'reference_ids': self.order_id.stock_reference_ids,
-        }
-        return values
-
-    def _launch_stock_rule_from_pos_order_lines(self):
-        procurements = []
-        for line in self:
-            line = line.with_company(line.company_id)
-            if line.product_id.type != 'consu':
-                continue
-
-            reference_ids = line.order_id.stock_reference_ids
-            if not reference_ids:
-                reference_ids = self.env['stock.reference'].create(line._prepare_reference_vals())
-                line.order_id.stock_reference_ids = [Command.set(reference_ids.ids)]
-
-            values = line._prepare_procurement_values()
-            product_qty = line.qty
-
-            procurement_uom = line.product_id.uom_id
-            procurements.append(self.env['stock.rule'].Procurement(
-                line.product_id, product_qty, procurement_uom,
-                line.order_id.partner_id.property_stock_customer,
-                line.name, line.order_id.name, line.order_id.company_id, values))
-        if procurements:
-            self.env['stock.rule'].run(procurements)
-
-        # This next block is currently needed only because the scheduler trigger is done by picking confirmation rather than stock.move confirmation
-        orders = self.mapped('order_id')
-        for order in orders:
-            pickings_to_confirm = order.picking_ids
-            if pickings_to_confirm:
-                # Trigger the Scheduler for Pickings
-                tracked_lines = order.lines.filtered(lambda l: l.product_id.tracking in ['lot', 'serial'])
-                lines_by_tracked_product = tracked_lines.grouped('product_id')
-                pickings_to_confirm.action_confirm()
-                for product, lines in lines_by_tracked_product.items():
-                    product_id = product.id
-                    moves = pickings_to_confirm.move_ids.filtered(lambda m: m.product_id.id == product_id)
-                    moves.move_line_ids.unlink()
-                    moves._add_mls_related_to_order(lines, are_qties_done=False)
-                    moves._recompute_state()
-        return True
-
-    def _is_product_storable_fifo_avco(self):
-        self.ensure_one()
-        return self.product_id.is_storable and self.product_id.cost_method in ['fifo', 'average']
-
-    def _get_product_cost_with_moves(self, moves):
-        self.ensure_one()
-        return moves._get_price_unit()
-
-    def _compute_total_cost(self, stock_moves):
+    def _compute_total_cost(self, at_closing=False):
         """
         Compute the total cost of the order lines.
-        :param stock_moves: recordset of `stock.move`, used for fifo/avco lines
         """
-        for line in self.filtered(lambda l: not l.is_total_cost_computed):
+        for line in self:
+            if line.is_total_cost_computed:
+                continue
             product = line.product_id
             cost_currency = product.sudo().cost_currency_id
-            moves = line._get_stock_moves_to_consider(stock_moves, product) if stock_moves else None
-            if moves and line._is_product_storable_fifo_avco():
-                product_cost = line._get_product_cost_with_moves(moves)
-                if cost_currency.is_zero(product_cost) and line.order_id.shipping_date:
-                    if line.refunded_orderline_id:
-                        product_cost = line.refunded_orderline_id.total_cost / line.refunded_orderline_id.qty
-                    else:
-                        product_cost = product.standard_price
-            else:
-                product_cost = product.standard_price
+            product_cost = line._get_product_cost(at_closing)
             line.total_cost = line.qty * cost_currency._convert(
                 from_amount=product_cost,
                 to_currency=line.currency_id,
@@ -344,10 +205,6 @@ class PosOrderLine(models.Model):
             )
             line.is_total_cost_computed = True
 
-    def _get_stock_moves_to_consider(self, stock_moves, product):
-        self.ensure_one()
-        return stock_moves.filtered(lambda ml: ml.product_id.id == product.id)
-
     @api.depends('price_subtotal', 'total_cost')
     def _compute_margin(self):
         for line in self:
@@ -413,3 +270,7 @@ class PosOrderLine(models.Model):
         self.ensure_one()
         original_price = self.tax_ids.compute_all(self.price_unit, self.currency_id, self.qty, product=self.product_id, partner=self.order_id.partner_id)['total_included']
         return original_price - self.price_subtotal_incl
+
+    def _get_product_cost(self, at_closing=False):
+        self.ensure_one()
+        return self.product_id.standard_price

--- a/addons/point_of_sale/models/pos_session.py
+++ b/addons/point_of_sale/models/pos_session.py
@@ -1,15 +1,14 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 from collections import defaultdict
 from datetime import timedelta
-from itertools import groupby, starmap
+from itertools import starmap
 from markupsafe import Markup
 import logging
 
 from odoo import api, fields, models, _
 from odoo.exceptions import AccessError, UserError, ValidationError
 from odoo.fields import Command, Domain
-from odoo.tools import float_is_zero, float_compare, frozendict, plaintext2html, split_every
-from odoo.tools.constants import PREFETCH_MAX
+from odoo.tools import float_is_zero, float_compare, frozendict, plaintext2html
 
 _logger = logging.getLogger(__name__)
 
@@ -78,9 +77,6 @@ class PosSession(models.Model):
     order_ids = fields.One2many('pos.order', 'session_id', string='Orders')
     order_count = fields.Integer(compute='_compute_order_count')
     statement_line_ids = fields.One2many('account.bank.statement.line', 'pos_session_id', string='Cash Lines', readonly=True)
-    failed_pickings = fields.Boolean(compute='_compute_picking_count')
-    picking_count = fields.Integer(compute='_compute_picking_count')
-    picking_ids = fields.One2many('stock.picking', 'pos_session_id')
     rescue = fields.Boolean(string='Recovery Session',
         help="Auto-generated session for orphan orders, ignored in constraints",
         readonly=True,
@@ -89,7 +85,6 @@ class PosSession(models.Model):
     payment_method_ids = fields.Many2many('pos.payment.method', related='config_id.payment_method_ids', string='Payment Methods')
     total_payments_amount = fields.Float(compute='_compute_total_payments_amount', string='Total Payments Amount')
     is_in_company_currency = fields.Boolean('Is Using Company Currency', compute='_compute_is_in_company_currency')
-    update_stock_at_closing = fields.Boolean('Stock should be updated at closing')
     bank_payment_ids = fields.One2many('account.payment', 'pos_session_id', 'Bank Payments', help='Account payments representing aggregated and bank split payments.')
 
     def write(self, vals):
@@ -137,11 +132,11 @@ class PosSession(models.Model):
 
     @api.model
     def _load_pos_data_models(self, config):
-        return ['pos.config', 'pos.preset', 'resource.calendar.attendance', 'pos.order', 'pos.order.line', 'pos.pack.operation.lot', 'pos.payment', 'pos.payment.method', 'pos.printer',
+        return ['pos.config', 'pos.preset', 'resource.calendar.attendance', 'pos.order', 'pos.order.line', 'pos.payment', 'pos.payment.method', 'pos.printer',
             'pos.category', 'pos.bill', 'res.company', 'account.tax', 'account.tax.group', 'product.template', 'product.product', 'product.attribute', 'product.attribute.custom.value',
             'product.template.attribute.line', 'product.template.attribute.value', 'product.combo', 'product.combo.item', 'res.users', 'res.partner', 'product.uom',
             'decimal.precision', 'uom.uom', 'res.country', 'res.country.state', 'res.lang', 'product.category', 'product.pricelist', 'product.pricelist.item',
-            'account.cash.rounding', 'account.fiscal.position', 'stock.picking.type', 'res.currency', 'pos.note', 'product.tag', 'ir.module.module', 'account.move', 'account.account', 'pos.product.template.snooze']
+            'account.cash.rounding', 'account.fiscal.position', 'res.currency', 'pos.note', 'product.tag', 'ir.module.module', 'account.move', 'account.account', 'pos.product.template.snooze']
 
     @api.model
     def _load_pos_data_domain(self, data, config):
@@ -151,7 +146,7 @@ class PosSession(models.Model):
     def _load_pos_data_fields(self, config):
         return [
             'id', 'name', 'user_id', 'config_id', 'start_at', 'stop_at',
-            'payment_method_ids', 'state', 'update_stock_at_closing', 'cash_register_balance_start', 'access_token'
+            'payment_method_ids', 'state', 'cash_register_balance_start', 'access_token'
         ]
 
     def load_data(self, models_to_load):
@@ -272,20 +267,6 @@ class PosSession(models.Model):
         for session in self:
             session.order_count = sessions_data.get(session.id, 0)
 
-    @api.depends('picking_ids', 'picking_ids.state')
-    def _compute_picking_count(self):
-        for session in self:
-            session.picking_count = self.env['stock.picking'].search_count([('pos_session_id', 'in', session.ids)])
-            session.failed_pickings = bool(self.env['stock.picking'].search([('pos_session_id', 'in', session.ids), ('state', '!=', 'done')], limit=1))
-
-    def action_stock_picking(self):
-        self.ensure_one()
-        action = self.env['ir.actions.act_window']._for_xml_id('stock.action_picking_tree_ready')
-        action['display_name'] = _('Pickings')
-        action['context'] = {}
-        action['domain'] = [('id', 'in', self.picking_ids.ids)]
-        return action
-
     @api.depends('cash_journal_id')
     def _compute_cash_control(self):
         # Only one cash register is supported by point_of_sale.
@@ -342,14 +323,8 @@ class PosSession(models.Model):
             # exists at the installation. If nothing is configured at the
             # installation we do the minimal configuration. Impossible to do in
             # the .xml files as the CoA is not yet installed.
-            pos_config = self.env['pos.config'].browse(config_id)
 
-            update_stock_at_closing = pos_config.company_id.point_of_sale_update_stock_quantities == "closing"
-
-            vals.update({
-                'config_id': config_id,
-                'update_stock_at_closing': update_stock_at_closing,
-            })
+            vals.update(self._get_default_session_vals(config_id))
 
         if self.env.user.has_group('point_of_sale.group_pos_user'):
             sessions = super(PosSession, self.sudo()).create(vals_list)
@@ -359,6 +334,10 @@ class PosSession(models.Model):
         sessions.action_pos_session_open()
         return sessions
 
+    @api.model
+    def _get_default_session_vals(self, config_id):
+        return {'config_id': config_id}
+
     def unlink(self):
         self.statement_line_ids.unlink()
         return super().unlink()
@@ -415,49 +394,14 @@ class PosSession(models.Model):
         return self._validate_session(balancing_account, amount_to_balance, bank_payment_method_diffs)
 
     def _validate_session(self, balancing_account=False, amount_to_balance=0, bank_payment_method_diffs=None):
-        bank_payment_method_diffs = bank_payment_method_diffs or {}
-        record = self.ensure_one()
-        if self.env.user.has_group('point_of_sale.group_pos_user'):
-            record = record.sudo()
-        data = {}
+        self.ensure_one()
         if self.get_session_orders().filtered(lambda o: o.state != 'cancel') or self.sudo().statement_line_ids:
             self.cash_real_transaction = sum(self.sudo().statement_line_ids.mapped('amount'))
             if self.state == 'closed':
                 raise UserError(_('This session is already closed.'))
             self._check_if_no_draft_orders()
             self._check_invoices_are_posted()
-            cash_difference_before_statements = self.cash_register_difference
-            if self.update_stock_at_closing:
-                self._create_picking_at_end_of_session()
-                self._get_closed_orders().filtered(lambda o: not o.is_total_cost_computed)._compute_total_cost_at_session_closing(self.picking_ids.move_ids)
-            # when the user is POS, update the record in sudo
-            data = record.with_company(record.company_id).with_context(
-                check_move_validity=False, skip_invoice_sync=True
-            )._create_account_move(balancing_account, amount_to_balance, bank_payment_method_diffs)
-
-            balance = sum(record.move_id.line_ids.mapped('balance'))
-            try:
-                with self.move_id._check_balanced({'records': self.move_id.sudo()}):
-                    pass
-            except UserError:
-                # Creating the account move is just part of a big database transaction
-                # when closing a session. There are other database changes that will happen
-                # before attempting to create the account move, such as, creating the picking
-                # records.
-                # We don't, however, want them to be committed when the account move creation
-                # failed; therefore, we need to roll back this transaction before showing the
-                # close session wizard.
-                self.env.cr.rollback()
-                return self._close_session_action(balance)
-
-            self.sudo()._post_statement_difference(cash_difference_before_statements)
-            if record.move_id.line_ids:
-                record.move_id.with_company(self.company_id)._post()
-                # Set the uninvoiced orders' state to 'done'
-                self.env['pos.order'].search([('session_id', '=', self.id), ('state', '=', 'paid')]).write({'state': 'done'})
-            else:
-                record.move_id.sudo().unlink()
-            self.sudo().with_company(self.company_id)._reconcile_account_move_lines(data)
+            self._process_session_validation(balancing_account, amount_to_balance, bank_payment_method_diffs)
         else:
             self.sudo()._post_statement_difference(self.cash_register_difference)
 
@@ -468,14 +412,43 @@ class PosSession(models.Model):
                     Markup("<br/><ul>%s</ul>") % Markup().join(Markup("<li>%s</li>") % order._get_html_link() for order in edited_orders)
                 )
                 self.message_post(body=body)
-
-        # Make sure to trigger reordering rules
-        self.picking_ids.move_ids.sudo()._trigger_scheduler()
-
         self.write({'state': 'closed'})
         self.env.flush_all()  # ensure sale.report is up to date
         return True
 
+    def _process_session_validation(self, balancing_account, amount_to_balance, bank_payment_method_diffs):
+        bank_payment_method_diffs = bank_payment_method_diffs or {}
+        record = self.ensure_one()
+        if self.env.user.has_group('point_of_sale.group_pos_user'):
+            record = record.sudo()
+        cash_difference_before_statements = self.cash_register_difference
+        # when the user is POS, update the record in sudo
+        data = record.with_company(record.company_id).with_context(
+            check_move_validity=False, skip_invoice_sync=True
+        )._create_account_move(balancing_account, amount_to_balance, bank_payment_method_diffs) or {}
+        balance = sum(record.move_id.line_ids.mapped('balance'))
+        try:
+            with self.move_id._check_balanced({'records': self.move_id.sudo()}):
+                pass
+        except UserError:
+            # Creating the account move is just part of a big database transaction
+            # when closing a session. There are other database changes that will happen
+            # before attempting to create the account move, such as, creating the picking
+            # records.
+            # We don't, however, want them to be committed when the account move creation
+            # failed; therefore, we need to roll back this transaction before showing the
+            # close session wizard.
+            self.env.cr.rollback()
+            return self._close_session_action(balance)
+        self.sudo()._post_statement_difference(cash_difference_before_statements)
+        if record.move_id.line_ids:
+            record.move_id.with_company(self.company_id)._post()
+            # Set the uninvoiced orders' state to 'done'
+            self.env['pos.order'].search([('session_id', '=', self.id), ('state', '=', 'paid')]).write({'state': 'done'})
+        else:
+            record.move_id.sudo().unlink()
+        self.sudo().with_company(self.company_id)._reconcile_account_move_lines(data)
+
     def _post_statement_difference(self, amount):
         if amount:
             if self.config_id.cash_control:
@@ -752,29 +725,6 @@ class PosSession(models.Model):
             'amount_authorized_diff': self.config_id.amount_authorized_diff if self.config_id.set_maximum_difference else None
         }
 
-    def _create_picking_at_end_of_session(self):
-        self.ensure_one()
-        lines_grouped_by_dest_location = {}
-        picking_type = self.config_id.picking_type_id
-
-        if not picking_type or not picking_type.default_location_dest_id:
-            session_destination_id = self.env['stock.warehouse']._get_partner_locations()[0].id
-        else:
-            session_destination_id = picking_type.default_location_dest_id.id
-
-        for order in self._get_closed_orders():
-            if order._force_create_picking_real_time() or order.shipping_date:
-                continue
-            destination_id = order.partner_id.property_stock_customer.id or session_destination_id
-            if destination_id in lines_grouped_by_dest_location:
-                lines_grouped_by_dest_location[destination_id] |= order.lines
-            else:
-                lines_grouped_by_dest_location[destination_id] = order.lines
-
-        for location_dest_id, lines in lines_grouped_by_dest_location.items():
-            pickings = self.env['stock.picking']._create_picking_from_pos_order_lines(location_dest_id, lines, picking_type)
-            pickings.write({'pos_session_id': self.id, 'origin': self.name})
-
     def _create_balancing_line(self, data, balancing_account, amount_to_balance):
         if not self.company_id.currency_id.is_zero(amount_to_balance):
             balancing_vals = self._prepare_balancing_line_vals(amount_to_balance, self.move_id, balancing_account)
@@ -817,7 +767,12 @@ class PosSession(models.Model):
             'ref': self.name,
         })
         self.write({'move_id': account_move.id})
+        data = self._get_account_move_data(bank_payment_method_diffs)
+        if balancing_account and amount_to_balance:
+            data = self._create_balancing_line(data, balancing_account, amount_to_balance)
+        return data
 
+    def _get_account_move_data(self, bank_payment_method_diffs):
         data = {'bank_payment_method_diffs': bank_payment_method_diffs or {}}
         data = self._accumulate_amounts(data)
         data = self._create_non_reconciliable_move_lines(data)
@@ -825,10 +780,6 @@ class PosSession(models.Model):
         data = self._create_pay_later_receivable_lines(data)
         data = self._create_cash_statement_lines_and_cash_move_lines(data)
         data = self._create_invoice_receivable_lines(data)
-        data = self._create_stock_valuation_lines(data)
-        if balancing_account and amount_to_balance:
-            data = self._create_balancing_line(data, balancing_account, amount_to_balance)
-
         return data
 
     def _accumulate_amounts(self, data):
@@ -850,9 +801,6 @@ class PosSession(models.Model):
         split_invoice_receivables = defaultdict(amounts)
         sales = defaultdict(amounts)
         taxes = defaultdict(tax_amounts)
-        stock_expense = defaultdict(amounts)
-        stock_return = defaultdict(amounts)
-        stock_valuation = defaultdict(amounts)
         rounding_difference = {'amount': 0.0, 'amount_converted': 0.0}
         # Track the receivable lines of the order's invoice payment moves for reconciliation
         # These receivable lines are reconciled to the corresponding invoice receivable lines
@@ -956,36 +904,11 @@ class PosSession(models.Model):
                 partners = (order.partner_id | order.partner_id.commercial_partner_id)
                 partners._increase_rank('customer_rank')
 
-        all_picking_ids = self.order_ids.filtered(lambda p: not p.is_invoiced and not p.shipping_date).picking_ids.ids + self.picking_ids.filtered(lambda p: not p.pos_order_id).ids
-        if all_picking_ids:
-            # Combine stock lines
-            stock_move_sudo = self.env['stock.move'].sudo()
-            stock_moves = stock_move_sudo.search([
-                ('picking_id', 'in', all_picking_ids),
-                ('product_id.valuation', '=', 'real_time'),
-                ('product_id.is_storable', '=', True),
-            ])
-            for stock_moves_batch in split_every(PREFETCH_MAX, stock_moves._ids, stock_moves.browse):
-                for move in stock_moves_batch:
-                    product_accounts = move.with_company(move.company_id).product_id._get_product_accounts()
-                    exp_key = product_accounts['expense']
-                    stock_key = product_accounts['stock_valuation']
-                    signed_product_qty = move.uom_id._compute_quantity(move.product_uom_qty, move.product_id.uom_id, round=False)
-                    if move._is_in():
-                        signed_product_qty *= -1
-                    amount = signed_product_qty * move._get_price_unit()
-                    stock_expense[exp_key] = self._update_amounts(stock_expense[exp_key], {'amount': amount}, move.picking_id.date_done, force_company_currency=True)
-                    if move._is_in():
-                        stock_return[stock_key] = self._update_amounts(stock_return[stock_key], {'amount': amount}, move.picking_id.date_done, force_company_currency=True)
-                    else:
-                        stock_valuation[stock_key] = self._update_amounts(stock_valuation[stock_key], {'amount': amount}, move.picking_id.date_done, force_company_currency=True)
-
         MoveLine = self.env['account.move.line'].with_context(check_move_validity=False, skip_invoice_sync=True)
 
         data.update({
             'taxes':                               taxes,
             'sales':                               sales,
-            'stock_expense':                       stock_expense,
             'split_receivables_bank':              split_receivables_bank,
             'combine_receivables_bank':            combine_receivables_bank,
             'split_receivables_cash':              split_receivables_cash,
@@ -993,8 +916,6 @@ class PosSession(models.Model):
             'combine_invoice_receivables':         combine_invoice_receivables,
             'split_receivables_pay_later':         split_receivables_pay_later,
             'combine_receivables_pay_later':       combine_receivables_pay_later,
-            'stock_return':                        stock_return,
-            'stock_valuation':                     stock_valuation,
             'combine_inv_payment_receivable_lines': combine_inv_payment_receivable_lines,
             'rounding_difference':                 rounding_difference,
             'MoveLine':                            MoveLine,
@@ -1007,13 +928,10 @@ class PosSession(models.Model):
         # Create account.move.line records for
         #   - sales
         #   - taxes
-        #   - stock expense
         #   - non-cash split receivables (not for automatic reconciliation)
         #   - non-cash combine receivables (not for automatic reconciliation)
         taxes = data.get('taxes')
         sales = data.get('sales')
-        stock_expense = data.get('stock_expense')
-        rounding_difference = data.get('rounding_difference')
         MoveLine = data.get('MoveLine')
 
         tax_vals = [self._get_tax_vals(key, amounts['amount'], amounts['amount_converted'], amounts['base_amount_converted']) for key, amounts in taxes.items()]
@@ -1025,19 +943,11 @@ class PosSession(models.Model):
                 'Please set corresponding tax account in each repartition line of the following taxes: \n%s',
                 ', '.join(tax_names_no_account)
             ))
-        rounding_vals = []
-
-        if not float_is_zero(rounding_difference['amount'], precision_rounding=self.currency_id.rounding) or not float_is_zero(rounding_difference['amount_converted'], precision_rounding=self.currency_id.rounding):
-            rounding_vals = [self._get_rounding_difference_vals(rounding_difference['amount'], rounding_difference['amount_converted'])]
 
         MoveLine.create(tax_vals)
         move_line_ids = MoveLine.create(list(starmap(self._get_sale_vals, sales.items())))
         for key, ml_id in zip(sales.keys(), move_line_ids.ids):
             sales[key]['move_line_id'] = ml_id
-        MoveLine.create(
-            [self._get_stock_expense_vals(key, amounts['amount'], amounts['amount_converted']) for key, amounts in stock_expense.items()]
-            + rounding_vals
-        )
 
         return data
 
@@ -1257,23 +1167,6 @@ class PosSession(models.Model):
         data.update({'split_invoice_receivable_lines': split_invoice_receivable_lines})
         return data
 
-    def _create_stock_valuation_lines(self, data):
-        MoveLine = data.get('MoveLine')
-        stock_valuation = data.get('stock_valuation')
-        stock_return = data.get('stock_return')
-
-        stock_valuation_vals = defaultdict(list)
-        stock_valuation_lines = {}
-        for stock_moves in [stock_valuation, stock_return]:
-            for account, amounts in stock_moves.items():
-                stock_valuation_vals[account].append(self._get_stock_valuation_vals(account, amounts['amount'], amounts['amount_converted']))
-
-        for stock_valuation_acc, vals in stock_valuation_vals.items():
-            stock_valuation_lines[stock_valuation_acc] = MoveLine.create(vals)
-
-        data.update({'stock_valuation_lines': stock_valuation_lines})
-        return data
-
     def _reconcile_account_move_lines(self, data):
         # reconcile cash receivable lines
         split_cash_statement_lines = data.get('split_cash_statement_lines')
@@ -1284,11 +1177,9 @@ class PosSession(models.Model):
         split_inv_payment_receivable_lines = data.get('split_inv_payment_receivable_lines')
         combine_invoice_receivable_lines = data.get('combine_invoice_receivable_lines')
         split_invoice_receivable_lines = data.get('split_invoice_receivable_lines')
-        stock_output_lines = data.get('stock_output_lines')
         payment_method_to_receivable_lines = data.get('payment_method_to_receivable_lines')
         payment_to_receivable_lines = data.get('payment_to_receivable_lines')
 
-
         all_lines = (
               split_cash_statement_lines
             | combine_cash_statement_lines
@@ -1318,20 +1209,6 @@ class PosSession(models.Model):
 
         return data
 
-    def _get_rounding_difference_vals(self, amount, amount_converted):
-        if self.config_id.cash_rounding:
-            partial_args = {
-                'name': 'Rounding line',
-                'move_id': self.move_id.id,
-            }
-            if float_compare(0.0, amount, precision_rounding=self.currency_id.rounding) > 0:    # loss
-                partial_args['account_id'] = self.config_id.rounding_method.loss_account_id.id
-                return self._debit_amounts(partial_args, -amount, -amount_converted)
-
-            if float_compare(0.0, amount, precision_rounding=self.currency_id.rounding) < 0:   # profit
-                partial_args['account_id'] = self.config_id.rounding_method.profit_account_id.id
-                return self._credit_amounts(partial_args, amount, amount_converted)
-
     def _get_split_receivable_vals(self, payment, amount, amount_converted):
         accounting_partner = self.env["res.partner"]._find_accounting_partner(payment.partner_id)
         if not accounting_partner:
@@ -1426,14 +1303,6 @@ class PosSession(models.Model):
             'balance': amount_converted,
         }
 
-    def _get_stock_expense_vals(self, exp_account, amount, amount_converted):
-        partial_args = {'account_id': exp_account.id, 'move_id': self.move_id.id}
-        return self._debit_amounts(partial_args, amount, amount_converted, force_company_currency=True)
-
-    def _get_stock_valuation_vals(self, stock_val_account, amount, amount_converted):
-        partial_args = {'account_id': stock_val_account.id, 'move_id': self.move_id.id}
-        return self._credit_amounts(partial_args, amount, amount_converted, force_company_currency=True)
-
     def _get_combine_statement_line_vals(self, journal, amount, payment_method):
         amount_values = self._prepare_statement_line_amount_values(journal, amount)
         return {
@@ -1623,15 +1492,13 @@ class PosSession(models.Model):
         return self.env['account.move.line'].search([('ref', 'in', diff_lines_ref + cost_move_lines)]).mapped('move_id')
 
     def _get_related_account_moves(self):
-        pickings = self.picking_ids | self._get_closed_orders().mapped('picking_ids')
         invoices = self.mapped('order_ids.account_move')
         invoice_payments = self.mapped('order_ids.payment_ids.account_move_id')
-        stock_account_moves = pickings.move_ids.account_move_id
         cash_moves = self.statement_line_ids.mapped('move_id')
         bank_payment_moves = self.bank_payment_ids.mapped('move_id')
         reversal_moves = self.mapped('order_ids.reversed_move_ids')
         other_related_moves = self._get_other_related_moves()
-        return invoices | invoice_payments | self.move_id | stock_account_moves | cash_moves | bank_payment_moves | reversal_moves | other_related_moves
+        return invoices | invoice_payments | self.move_id | cash_moves | bank_payment_moves | reversal_moves | other_related_moves
 
     def _get_receivable_account(self, payment_method):
         """Returns the default pos receivable account if no receivable_account_id is set on the payment method."""
@@ -1824,18 +1691,3 @@ class PosSession(models.Model):
 
     def _get_closed_orders(self):
         return self.order_ids.filtered(lambda o: o.state not in ['draft', 'cancel'])
-
-
-class StockRule(models.Model):
-    _inherit = 'stock.rule'
-
-    @api.model
-    def _run_scheduler_tasks(self, use_new_cursor=False, company_id=False):
-        super()._run_scheduler_tasks(use_new_cursor=use_new_cursor, company_id=company_id)
-        self.env['pos.session']._alert_old_session()
-        if use_new_cursor:
-            self.env['ir.cron']._commit_progress(1)
-
-    @api.model
-    def _get_scheduler_tasks_to_do(self):
-        return super()._get_scheduler_tasks_to_do() + 1

--- a/addons/point_of_sale/models/product_template.py
+++ b/addons/point_of_sale/models/product_template.py
@@ -170,7 +170,7 @@ class ProductTemplate(models.Model):
     def _load_pos_data_fields(self, config_id):
         return [
             'id', 'display_name', 'standard_price', 'categ_id', 'pos_categ_ids', 'taxes_id', 'barcode', 'name', 'list_price', 'is_favorite',
-            'default_code', 'to_weight', 'uom_id', 'description_sale', 'description', 'tracking', 'type', 'service_tracking', 'is_storable',
+            'default_code', 'to_weight', 'uom_id', 'description_sale', 'description', 'type', 'service_tracking', 'is_storable',
             'write_date', 'color', 'pos_sequence', 'available_in_pos', 'attribute_line_ids', 'active', 'image_128', 'combo_ids', 'product_variant_ids', 'public_description',
             'pos_optional_product_ids', 'sequence', 'product_tag_ids'
         ]
@@ -181,29 +181,7 @@ class ProductTemplate(models.Model):
         pos_limited_loading = self.env.context.get('pos_limited_loading', True)
         if limit_count and pos_limited_loading:
             query = self._search(self._load_pos_data_domain(data, config), bypass_access=True)
-            sql = SQL(
-                """
-                    WITH pm AS (
-                        SELECT pp.product_tmpl_id,
-                            MAX(sml.write_date) date
-                        FROM stock_move_line sml
-                        JOIN product_product pp ON sml.product_id = pp.id
-                        GROUP BY pp.product_tmpl_id
-                    )
-                    SELECT product_template.id
-                        FROM %s
-                    LEFT JOIN pm ON product_template.id = pm.product_tmpl_id
-                        WHERE %s
-                    ORDER BY product_template.is_favorite DESC NULLS LAST,
-                        CASE WHEN product_template.type = 'service' THEN 1 ELSE 0 END DESC,
-                        pm.date DESC NULLS LAST,
-                        product_template.write_date DESC
-                    LIMIT %s
-                """,
-                query.from_clause,
-                query.where_clause or SQL("TRUE"),
-                limit_count,
-            )
+            sql = self._get_load_product_template_sql(query, limit_count)
             product_tmpl_ids = [r[0] for r in self.env.execute_query(sql)]
             products = self._load_product_with_domain([('id', 'in', product_tmpl_ids)])
         else:
@@ -386,27 +364,6 @@ class ProductTemplate(models.Model):
         price_per_pricelist_id = pricelists._price_get(template_or_variant, quantity) if pricelists else False
         pricelist_list = [{'name': pl.name, 'price': price_per_pricelist_id[pl.id]} for pl in pricelists]
 
-        # Warehouses
-        warehouse_list = [
-            {'id': w.id,
-            'name': w.name,
-            'available_quantity': template_or_variant.with_context({'warehouse_id': w.id}).qty_available,
-            'free_qty': (
-                    template_or_variant.with_context({'warehouse_id': w.id}).free_qty
-                    if product_variant
-                    else sum(self.product_variant_ids.with_context({'warehouse_id': w.id}).mapped('free_qty'))
-                ),
-            'forecasted_quantity': template_or_variant.with_context({'warehouse_id': w.id}).virtual_available,
-            'uom': template_or_variant.uom_name}
-            for w in self.env['stock.warehouse'].search([('company_id', '=', config.company_id.id)])]
-
-        if config.picking_type_id.warehouse_id:
-            # Sort the warehouse_list, prioritizing config.picking_type_id.warehouse_id
-            warehouse_list = sorted(
-                warehouse_list,
-                key=lambda w: w['id'] != config.picking_type_id.warehouse_id.id
-            )
-
         # Suppliers
         supplier_list = []
         for group in self.seller_ids.grouped('partner_id').values():
@@ -428,8 +385,24 @@ class ProductTemplate(models.Model):
         return {
             'all_prices': all_prices,
             'pricelists': pricelist_list,
-            'warehouses': warehouse_list,
             'suppliers': supplier_list,
             'variants': variant_list,
             'optional_products': self.pos_optional_product_ids.read(['id', 'name', 'list_price']),
+            'free_qty': template_or_variant.qty_available,
+            'uom': template_or_variant.uom_name,
         }
+
+    def _get_load_product_template_sql(self, query, limit_count):
+        return SQL("""
+            SELECT product_template.id
+                FROM %s
+                WHERE %s
+            ORDER BY product_template.is_favorite DESC NULLS LAST,
+                CASE WHEN product_template.type = 'service' THEN 1 ELSE 0 END DESC,
+                product_template.write_date DESC
+            LIMIT %s
+            """,
+            query.from_clause,
+            query.where_clause or SQL("TRUE"),
+            limit_count,
+        )

--- a/addons/point_of_sale/models/res_company.py
+++ b/addons/point_of_sale/models/res_company.py
@@ -7,11 +7,6 @@ class ResCompany(models.Model):
     _name = 'res.company'
     _inherit = ['res.company', 'pos.load.mixin']
 
-    point_of_sale_update_stock_quantities = fields.Selection([
-            ('closing', 'At the session closing'),
-            ('real', 'In real time'),
-            ], default='real', string="Update quantities in stock",
-            help="At the session closing: A picking is created for the entire session when it's closed\n In real time: Each order sent to the server create its own picking")
     point_of_sale_use_ticket_qr_code = fields.Boolean(
         string='Self-service invoicing',
         default=True,

--- a/addons/point_of_sale/models/res_config_settings.py
+++ b/addons/point_of_sale/models/res_config_settings.py
@@ -39,7 +39,6 @@ class ResConfigSettings(models.TransientModel):
     module_pos_qfpay = fields.Boolean(string="QFPay Payment Terminal", help="The transactions are processed by QFPay. Set your QFPay credentials on the related payment method.")
     module_pos_dpopay = fields.Boolean(string="DPO Pay Payment Terminal", help="The transactions are processed by DPO Pay. Set your DPO Pay credentials on the related payment method.")
     module_pos_pricer = fields.Boolean(string="Pricer electronic price tags", help="Display the price of your products through electronic price tags")
-    update_stock_quantities = fields.Selection(related="company_id.point_of_sale_update_stock_quantities", readonly=False)
     account_default_pos_receivable_account_id = fields.Many2one(string='Default Account Receivable (PoS)', related='company_id.account_default_pos_receivable_account_id', readonly=False, check_company=True)
     barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', related='company_id.nomenclature_id', readonly=False)
     use_kiosk_mode = fields.Boolean(string="Is Kiosk Mode", default=False)
@@ -98,17 +97,13 @@ class ResConfigSettings(models.TransientModel):
     pos_other_devices = fields.Boolean(related='pos_config_id.other_devices', readonly=False)
     pos_preparation_devices = fields.Boolean(related='pos_config_id.preparation_devices', readonly=False)
     pos_payment_method_ids = fields.Many2many(related='pos_config_id.payment_method_ids', readonly=False)
-    pos_picking_policy = fields.Selection(related='pos_config_id.picking_policy', readonly=False)
-    pos_picking_type_id = fields.Many2one(related='pos_config_id.picking_type_id', readonly=False)
     pos_pricelist_id = fields.Many2one('product.pricelist', string='Default Pricelist', compute='_compute_pos_pricelist_id', readonly=False, store=True)
     pos_receipt_footer = fields.Text(string='Receipt Footer', compute='_compute_pos_receipt_header_footer', readonly=False, store=True)
     pos_receipt_header = fields.Text(string='Receipt Header', compute='_compute_pos_receipt_header_footer', readonly=False, store=True)
     pos_restrict_price_control = fields.Boolean(related='pos_config_id.restrict_price_control', readonly=False)
     pos_rounding_method = fields.Many2one(related='pos_config_id.rounding_method', readonly=False)
-    pos_route_id = fields.Many2one(related='pos_config_id.route_id', readonly=False)
     pos_selectable_categ_ids = fields.Many2many('pos.category', compute='_compute_pos_selectable_categ_ids')
     pos_set_maximum_difference = fields.Boolean(related='pos_config_id.set_maximum_difference', readonly=False)
-    pos_ship_later = fields.Boolean(related='pos_config_id.ship_later', readonly=False)
     pos_tax_regime_selection = fields.Boolean(related='pos_config_id.tax_regime_selection', readonly=False)
     pos_tip_product_id = fields.Many2one(related='pos_config_id.tip_product_id', readonly=False, store=True)
     pos_set_tip_after_payment = fields.Boolean(related='pos_config_id.set_tip_after_payment', store=True, readonly=False)
@@ -116,7 +111,6 @@ class ResConfigSettings(models.TransientModel):
     pos_tip_percentage_2 = fields.Integer(related='pos_config_id.tip_percentage_2', readonly=False)
     pos_tip_percentage_3 = fields.Integer(related='pos_config_id.tip_percentage_3', readonly=False)
     pos_use_pricelist = fields.Boolean(related='pos_config_id.use_pricelist', readonly=False)
-    pos_warehouse_id = fields.Many2one(related='pos_config_id.warehouse_id', readonly=False, string="Warehouse (PoS)")
     point_of_sale_use_ticket_qr_code = fields.Boolean(related='company_id.point_of_sale_use_ticket_qr_code', readonly=False)
     pos_auto_validate_electronic_payment = fields.Boolean(related='pos_config_id.auto_validate_electronic_payment', readonly=False, string="Automatically validates orders paid with an electronic payment (terminals, qr codes, ...).")
     pos_trusted_config_ids = fields.Many2many(related='pos_config_id.trusted_config_ids', readonly=False, domain="[('id', '!=', pos_config_id), ('module_pos_restaurant', '=', False)]")

--- a/addons/point_of_sale/models/stock_picking.py
+++ b/None
@@ -1,297 +0,0 @@
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import api, fields, models, _
-from odoo.exceptions import UserError, ValidationError
-
-
-class StockPicking(models.Model):
-    _inherit = 'stock.picking'
-
-    pos_session_id = fields.Many2one('pos.session', index=True)
-    pos_order_id = fields.Many2one('pos.order', index=True)
-
-    def _prepare_picking_vals(self, partner, picking_type, location_id, location_dest_id):
-        return {
-            'partner_id': partner.id if partner else False,
-            'user_id': False,
-            'picking_type_id': picking_type.id,
-            'move_type': 'direct',
-            'location_id': location_id,
-            'location_dest_id': location_dest_id,
-            'state': 'draft',
-        }
-
-
-    @api.model
-    def _create_picking_from_pos_order_lines(self, location_dest_id, lines, picking_type, partner=False):
-        """We'll create some picking based on order_lines"""
-
-        pickings = self.env['stock.picking']
-        stockable_lines = lines.filtered(lambda l: l.product_id.type == 'consu' and not l.product_id.uom_id.is_zero(l.qty))
-        if not stockable_lines:
-            return pickings
-        positive_lines = stockable_lines.filtered(lambda l: l.qty > 0)
-        negative_lines = stockable_lines - positive_lines
-
-        if positive_lines:
-            location_id = picking_type.default_location_src_id.id
-            positive_picking = self.env['stock.picking'].create(
-                self._prepare_picking_vals(partner, picking_type, location_id, location_dest_id)
-            )
-
-            positive_picking._create_move_from_pos_order_lines(positive_lines)
-            self.env.flush_all()
-            try:
-                with self.env.cr.savepoint():
-                    positive_picking._action_done()
-            except (UserError, ValidationError):
-                pass
-
-            pickings |= positive_picking
-        if negative_lines:
-            if picking_type.return_picking_type_id:
-                return_picking_type = picking_type.return_picking_type_id
-                return_location_id = return_picking_type.default_location_dest_id.id
-            else:
-                return_picking_type = picking_type
-                return_location_id = picking_type.default_location_src_id.id
-
-            negative_picking = self.env['stock.picking'].create(
-                self._prepare_picking_vals(partner, return_picking_type, location_dest_id, return_location_id)
-            )
-            negative_picking._create_move_from_pos_order_lines(negative_lines)
-            self.env.flush_all()
-            try:
-                with self.env.cr.savepoint():
-                    negative_picking._action_done()
-            except (UserError, ValidationError):
-                pass
-            pickings |= negative_picking
-        return pickings
-
-    def _prepare_stock_move_vals(self, first_line, order_lines):
-        return {
-            'uom_id': first_line.product_id.uom_id.id,
-            'picking_id': self.id,
-            'picking_type_id': self.picking_type_id.id,
-            'product_id': first_line.product_id.id,
-            'product_uom_qty': abs(sum(order_lines.mapped('qty'))),
-            'location_id': self.location_id.id,
-            'location_dest_id': self.location_dest_id.id,
-            'company_id': self.company_id.id,
-            'never_product_template_attribute_value_ids': first_line.attribute_value_ids.filtered(lambda a: a.attribute_id.create_variant == 'no_variant'),
-        }
-
-    def _create_move_from_pos_order_lines(self, lines):
-        self.ensure_one()
-
-        def get_grouping_key(line):
-            return (line.product_id.id, tuple(sorted(line.attribute_value_ids.ids)))
-
-        move_vals = [
-            self._prepare_stock_move_vals(order_lines[0], order_lines)
-            for order_lines in lines.grouped(get_grouping_key).values()
-        ]
-        moves = self.env['stock.move'].create(move_vals)
-        confirmed_moves = moves._action_confirm()
-        confirmed_moves._add_mls_related_to_order(lines, are_qties_done=True)
-        confirmed_moves.picked = True
-        self._link_owner_on_return_picking(lines)
-
-    def _link_owner_on_return_picking(self, lines):
-        """This method tries to retrieve the owner of the returned product"""
-        if lines and lines[0].order_id.refunded_order_id.picking_ids:
-            returned_lines_picking = lines[0].order_id.refunded_order_id.picking_ids
-            returnable_qty_by_product = {}
-            for move_line in returned_lines_picking.move_line_ids:
-                returnable_qty_by_product[(move_line.product_id.id, move_line.owner_id.id or 0)] = move_line.quantity
-            for move in self.move_line_ids:
-                for keys in returnable_qty_by_product:
-                    if move.product_id.id == keys[0] and keys[1] and returnable_qty_by_product[keys] > 0:
-                        move.write({'owner_id': keys[1]})
-                        returnable_qty_by_product[keys] -= move.quantity
-
-
-    def _send_confirmation_email(self):
-        # Avoid sending Mail/SMS for POS deliveries
-        pickings = self.filtered(lambda p: p.picking_type_id != p.picking_type_id.warehouse_id.pos_type_id)
-        return super(StockPicking, pickings)._send_confirmation_email()
-
-
-class StockPickingType(models.Model):
-    _name = 'stock.picking.type'
-    _inherit = ['stock.picking.type', 'pos.load.mixin']
-
-    @api.depends('warehouse_id')
-    def _compute_hide_reservation_method(self):
-        super()._compute_hide_reservation_method()
-        for picking_type in self:
-            if picking_type == picking_type.warehouse_id.pos_type_id:
-                picking_type.hide_reservation_method = True
-
-    @api.constrains('active')
-    def _check_active(self):
-        for picking_type in self:
-            if picking_type.active:
-                continue
-            pos_config = self.env['pos.config'].sudo().search([('picking_type_id', '=', picking_type.id)], limit=1)
-            if pos_config:
-                raise ValidationError(_("You cannot archive '%(picking_type)s' as it is used by POS configuration '%(config)s'.", picking_type=picking_type.name, config=pos_config.name))
-
-    @api.model
-    def _load_pos_data_domain(self, data, config):
-        return [('id', '=', config.picking_type_id.id)]
-
-    @api.model
-    def _load_pos_data_fields(self, config):
-        return ['id', 'use_create_lots', 'use_existing_lots']
-
-
-class StockMove(models.Model):
-    _inherit = 'stock.move'
-
-    def _get_new_picking_values(self):
-        vals = super()._get_new_picking_values()
-        orders = self.reference_ids.pos_order_ids
-        if orders:
-            order = orders.filtered(lambda o: o.is_refund and o.state == 'paid')[:1] or orders[:1]
-            vals['pos_session_id'] = order.session_id.id
-            vals['pos_order_id'] = order.id
-        return vals
-
-    def _key_assign_picking(self):
-        keys = super(StockMove, self)._key_assign_picking()
-        return keys + (self.reference_ids.pos_order_ids,)
-
-    @api.model
-    def _prepare_lines_data_dict(self, order_lines):
-        return {
-            product.id: {'order_lines': lines}
-            for product, lines in order_lines.grouped('product_id').items()
-        }
-
-    def _create_production_lots_for_pos_order(self, lines):
-        ''' Search for existing lots and create missing ones.
-
-            :param lines: pos order lines with pack lot ids.
-            :type lines: pos.order.line recordset.
-
-            :return stock.lot recordset.
-        '''
-        valid_lots = self.env['stock.lot']
-        moves = self.filtered(lambda m: m.picking_type_id.use_existing_lots)
-        # Already called in self._action_confirm() but just to be safe when coming from _launch_stock_rule_from_pos_order_lines.
-        self._check_company()
-        if moves:
-            moves_product_ids = set(moves.mapped('product_id').ids)
-            lots = lines.pack_lot_ids.filtered(lambda l: l.lot_name and l.product_id.id in moves_product_ids)
-            lots_data = set(lots.mapped(lambda l: (l.product_id.id, l.lot_name)))
-            existing_lots = self.env['stock.lot'].search([
-                '|', ('company_id', '=', False), ('company_id', '=', moves[0].picking_type_id.company_id.id),
-                ('product_id', 'in', lines.product_id.ids),
-                ('name', 'in', lots.mapped('lot_name')),
-            ])
-            #The previous search may return (product_id.id, lot_name) combinations that have no matching in lines.pack_lot_ids.
-            for lot in existing_lots:
-                if (lot.product_id.id, lot.name) in lots_data:
-                    valid_lots |= lot
-                    lots_data.remove((lot.product_id.id, lot.name))
-            moves = moves.filtered(lambda m: m.picking_type_id.use_create_lots)
-            if moves:
-                moves_product_ids = set(moves.mapped('product_id').ids)
-                missing_lot_values = []
-                for lot_product_id, lot_name in filter(lambda l: l[0] in moves_product_ids, lots_data):
-                    missing_lot_values.append({'company_id': self.company_id.id, 'product_id': lot_product_id, 'name': lot_name})
-                valid_lots |= self.env['stock.lot'].create(missing_lot_values)
-        return valid_lots
-
-    def _add_mls_related_to_order(self, related_order_lines, are_qties_done=True):
-        lines_data = self._prepare_lines_data_dict(related_order_lines)
-        # Moves with product_id not in related_order_lines. This can happend e.g. when product_id has a phantom-type bom.
-        moves_to_assign = self.filtered(lambda m: m.product_id.id not in lines_data or m.product_id.tracking not in ['lot', 'serial']
-                                                  or (not m.picking_type_id.use_existing_lots and not m.picking_type_id.use_create_lots))
-
-        # Check for any conversion issues in the moves before setting quantities
-        uoms_with_issues = set()
-        for move in moves_to_assign.filtered(lambda m: m.product_uom_qty and m.uom_id != m.product_id.uom_id):
-            converted_qty = move.uom_id._compute_quantity(
-                move.product_uom_qty,
-                move.product_id.uom_id,
-                rounding_method='HALF-UP'
-            )
-            if not converted_qty:
-                uoms_with_issues.add(
-                    (move.uom_id.name, move.product_id.uom_id.name)
-                )
-
-        if uoms_with_issues:
-            error_message_lines = [
-                _("Conversion Error: The following unit of measure conversions result in a zero quantity due to rounding:")
-            ]
-            for uom_from, uom_to in uoms_with_issues:
-                error_message_lines.append(_(' - From "%(uom_from)s" to "%(uom_to)s"', uom_from=uom_from, uom_to=uom_to))
-
-            error_message_lines.append(
-                _("\nThis issue occurs because the quantity becomes zero after rounding during the conversion. "
-                "To fix this, adjust the conversion factors or rounding method to ensure that even the smallest quantity in the original unit "
-                "does not round down to zero in the target unit.")
-            )
-
-            raise UserError('\n'.join(error_message_lines))
-
-        for move in moves_to_assign:
-            move.quantity = move.product_uom_qty
-        moves_remaining = self - moves_to_assign
-        existing_lots = moves_remaining._create_production_lots_for_pos_order(related_order_lines)
-        move_lines_to_create = []
-        if are_qties_done:
-            for move in moves_remaining:
-                move.move_line_ids.unlink()
-                for line in lines_data[move.product_id.id]['order_lines']:
-                    for lot in line.pack_lot_ids.filtered(lambda l: l.lot_name):
-                        qty = 1 if line.product_id.tracking == 'serial' else abs(line.qty)
-                        if existing_lots:
-                            existing_lot = existing_lots.filtered_domain([('product_id', '=', line.product_id.id), ('name', '=', lot.lot_name)])
-                            quants = self.env['stock.quant']
-                            if existing_lot:
-                                quants = self.env['stock.quant'].search(
-                                    [('lot_id', '=', existing_lot.id), ('quantity', '>', '0.0'), ('location_id', 'child_of', move.location_id.id)],
-                                    order='id desc',
-                                )
-                            qty_left_to_assign = qty
-                            for quant in quants:
-                                if qty_left_to_assign <= 0:
-                                    break
-                                qty_chg = min(qty_left_to_assign, quant.quantity)
-                                ml_vals = dict(move._prepare_move_line_vals(qty_chg))
-                                qty_left_to_assign -= qty_chg
-                                ml_vals.update({
-                                    'quant_id': quant.id,
-                                })
-                                move_lines_to_create.append(ml_vals)
-                            if qty_left_to_assign > 0:
-                                ml_vals = dict(move._prepare_move_line_vals(qty_left_to_assign))
-                                ml_vals.update({
-                                    'lot_name': existing_lot.name,
-                                    'lot_id': existing_lot.id,
-                                })
-                                move_lines_to_create.append(ml_vals)
-                        else:
-                            ml_vals = dict(move._prepare_move_line_vals(qty))
-                            ml_vals.update({'lot_name': lot.lot_name})
-                            move_lines_to_create.append(ml_vals)
-
-            self.env['stock.move.line'].create(move_lines_to_create)
-        else:
-            for move in moves_remaining:
-                for line in lines_data[move.product_id.id]['order_lines']:
-                    for lot in line.pack_lot_ids.filtered(lambda l: l.lot_name):
-                        if line.product_id.tracking == 'serial':
-                            qty = 1
-                        else:
-                            qty = abs(line.qty)
-                        if existing_lots:
-                            existing_lot = existing_lots.filtered_domain([('product_id', '=', line.product_id.id), ('name', '=', lot.lot_name)])
-                            if existing_lot:
-                                move._update_reserved_quantity(qty, move.location_id, lot_id=existing_lot)
-                                continue

--- a/addons/point_of_sale/models/stock_reference.py
+++ b/None
@@ -1,9 +0,0 @@
-from odoo import fields, models
-
-
-class StockReference(models.Model):
-    _inherit = 'stock.reference'
-
-    pos_order_ids = fields.Many2many(
-        'pos.order', 'stock_reference_pos_order_rel', 'reference_id',
-        'pos_order_id', string="PoS Orders")

--- a/addons/point_of_sale/models/stock_warehouse.py
+++ b/None
@@ -1,49 +0,0 @@
-# -*- coding: utf-8 -*-
-
-from odoo import models, fields, api, _
-
-
-class StockWarehouse(models.Model):
-    _inherit = "stock.warehouse"
-
-    pos_type_id = fields.Many2one('stock.picking.type', string="Point of Sale Operation Type", copy=False)
-
-    def _get_sequence_values(self, name=False, code=False):
-        sequence_values = super()._get_sequence_values(name=name, code=code)
-        sequence_values.update({
-            'pos_type_id': {
-                'name': _('%(name)s Picking POS', name=self.name),
-                'prefix': self.code + '/POS/',
-                'padding': 5,
-                'company_id': self.company_id.id,
-            }
-        })
-        return sequence_values
-
-    def _get_picking_type_update_values(self):
-        picking_type_update_values = super()._get_picking_type_update_values()
-        picking_type_update_values.update({
-            'pos_type_id': {'default_location_src_id': self.lot_stock_id.id}
-        })
-        return picking_type_update_values
-
-    def _get_picking_type_create_values(self, max_sequence):
-        picking_type_create_values, max_sequence = super()._get_picking_type_create_values(max_sequence)
-        picking_type_create_values.update({
-            'pos_type_id': {
-                'name': _('PoS Orders'),
-                'code': 'outgoing',
-                'default_location_src_id': self.lot_stock_id.id,
-                'default_location_dest_id': self.env.ref('stock.stock_location_customers').id,
-                'sequence': max_sequence + 1,
-                'company_id': self.company_id.id,
-            }
-        })
-        return picking_type_create_values, max_sequence + 2
-
-    @api.model
-    def _create_missing_pos_picking_types(self):
-        warehouses = self.env['stock.warehouse'].search([('pos_type_id', '=', False)])
-        for warehouse in warehouses:
-            new_vals = warehouse._create_or_update_sequences_and_picking_types()
-            warehouse.write(new_vals)
